kythe/docs/rfc/2967.md

```diff
@@ -0,0 +1,64 @@
+
+
+## Overview {#overview}
+
+Rather than introduce any new VName fields, nodes or edges build configuration can be adequately represented via a signature-relevant fact on anchor nodes and some ancillary clarifications to the guidelines for root and corpus.  This is predicated on:
+
+
+1.  Non-generated source files exist independent of build configuration.
+1.  Nodes must already incorporate relevant information into their signature.  This includes any elements related to build-configuration which materially alters the represented entity. This is currently already done by non-anchor (aka "semantic") nodes, but build configuration has not been considered relevant for anchors.
+1.  Certain languages will necessarily emit different anchors under different build configurations (C/C++ are particularly subject to this).
+1.  Generated source files must incorporate the build configuration into root in some fashion, if relevant.
+1.  The relevant information should be in a form which is readily available for filtering and display.
+
+The end result is that build configuration is most tightly bound to anchors both semantically and practically.  This design also allows for including references to entities regardless of build configuration as well as filtering only references for a particular set.  Indexers for which build configuration is irrelevant may omit the fact.
+
+
+## Detailed Design {#detailed-design}
+
+
+### Graph {#graph}
+
+The Kythe graph will be augmented such that:
+
+
+1.  File nodes with a non-empty root VName field are explicitly considered to have been generated by the build.  If the contents of generated files differ with the build configuration, the root must include a unique string per build configuration.  This follows from the general requirements that unique graph entities be given unique VNames.  This allows clients to trivially know if they should request the source for a particular file from the underlying repository or from Kythe.
+1.  Corpus represents a stable name to a particular logical body of code.  Different corpus names may map to different parts of the same underlying repository.  That mapping resides outside the Kythe graph[^1].  In particular, this can be used to accommodate source code which resides in a particularly repository, but is logically distinct from the primary corpus (perhaps due to using a different build system).  For example, within Piper the primary corpus is "google3" but the source code for several languages' standard library resides in third_party.  Presently, Java and C++ use distinct corpora for these references, but Go uses a separate root.  Go will be changed to use the "gostdlib" corpus for this.
```

Comment:
Go only uses the root this way inside Google, because of how Blaze is organized. It doesn't use the root in the "native" Go implementation. 

cf. https://github.com/kythe/kythe/blob/c453a319b76da635c7915c8376d0121118c2d8bd/kythe/go/extractors/govname/govname.go#L34

Reply:
yes, this is probably more relevant internally but is included for completeness and reference